src/compiler/checker.ts

```diff
@@ -31399,6 +31404,70 @@ namespace ts {
                 getNonNullableType(checkExpression(node.expression));
         }
 
+        function checkExpressionWithTypeArguments(node: ExpressionWithTypeArguments | TypeQueryNode) {
+            checkGrammarExpressionWithTypeArguments(node);
+            const exprType = node.kind === SyntaxKind.ExpressionWithTypeArguments ? checkExpression(node.expression) :
+                isThisIdentifier(node.exprName) ? checkThisExpression(node.exprName) :
+                checkExpression(node.exprName);
+            const typeArguments = node.typeArguments;
+            if (!some(typeArguments)) {
+                return exprType;
+            }
+            const funcType = getApparentType(exprType);
+            if (funcType === silentNeverType || isErrorType(funcType)) {
+                return funcType;
+            }
+            let hasSomeApplicableSignature = false;
+            let nonApplicableType: Type | undefined;
+            let hasSignatures: boolean;
+            let hasApplicableSignature: boolean;
+            const result = mapType(funcType, getInstantiatedType);
+            const errorType = hasSomeApplicableSignature ? nonApplicableType : funcType;
+            if (errorType) {
+                diagnostics.add(createDiagnosticForNodeArray(getSourceFileOfNode(node), typeArguments, Diagnostics.Type_0_has_no_signatures_for_which_the_type_argument_list_is_applicable, typeToString(errorType)));
+            }
+            return result;
+
+            function getInstantiatedType(type: Type): Type {
+                hasSignatures = false;
+                hasApplicableSignature = false;
+                const result = getInstantiatedTypeWorker(type);
+                hasSomeApplicableSignature ||= hasApplicableSignature;
+                if (hasSignatures && !hasApplicableSignature) {
+                    nonApplicableType ??= type;
+                }
+                return result;
+            }
+
+            function getInstantiatedTypeWorker(type: Type): Type {

```

Comment:
I feel like there should probably also be a deferred version that is instantiated for indexed accesses or type parameters, since if a type parameter is constrained to a thing-with-signatures-with-type-arguments it's arguably reasonable to say you want to instantiate them (later).

Also, not looking into conditionals feels a bit odd - shouldn't it instantiate the conditional, but with the true and false branch types instantiated?

Reply:
See my comment above regarding instantiable types. As we discussed yesterday, it is possible to imagine a deferred type constructor that applies a given type argument list to a given type, but this would (a) require new type syntax that is distinct from existing generic type reference syntax and (b) add a significant amount of complexity to type relations, type instantiation, and type inference. Unless there are some really compelling real world scenarios, I'd rather not go there.

Reply:
@ahejlsberg I believe a notable real-world use case for what you're describing is generators, as correctly typing them requires specifying a type-level function to derive a `yield` expression's result type from its operand type. #32523, essentially. That would improve the generator experience significantly for libraries like `co`, `redux-saga`, etc.

I'd personally love to see such a feature at some point, but I agree that it's a tougher sell on the complexity side.

Reply:
@treybrisbane Actually, I was talking about some new type constructor `F<<T>>` that would represent applying the type argument list `T` to the type `F` (totally made up syntax here). Effectively a deferred version of what this PR implements. What you're talking about, I think, is higher-kinded types, i.e. the ability to have generic types take other generic types as type arguments. That's an even more complex feature request, discussed in #1213.

Reply:
Ahh, my bad! Yeah I slightly misinterpreted your comment. 😅

Apologies!